### PR TITLE
DRILL-3623: For limit 0 queries, use a shorter path when result column types are known

### PR DESCRIPTION
**Pull request by sudheeshkatkam**
Moving from #193 to here.

+ There is a pull request open for first commit (DRILL-4372: #397).
+ Second commit has a "nice to have" check: ensuring planning and execution types patch.
+ My changes are in the third commit (e4cfdfa). Please review this.

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/DirectScanPrule.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * limitations under the License.
+ */
+package org.apache.drill.exec.planner.physical;
+
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.drill.exec.planner.logical.DrillDirectScanRel;
+import org.apache.drill.exec.planner.logical.RelOptHelper;
+
```
Comment (hsuanyi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/DirectScanPrule.java]: java doc?

File: exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/limit/TestLimit0.java [hunk 2]
```diff
@@ -0,0 +1,677 @@
...
+        .sqlQuery(wrapLimit0(query))
+        .schemaBaseLine(expectedSchema)
+        .go();
+
+    checkThatQueryPlanIsOptimized(query);
+
+  }
+
+  @Test // negative aggregation test case
+  public void rank() throws Exception {
```
Comment (StevenMPhillips) [inline exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/limit/TestLimit0.java]: Why is this query not optimized?

**Comment by hsuanyi** [comment]
Except for my comments (which are for future improvements and can be filed as follow-up jira issues)
LGTM +1 (non-binding)

**Comment by StevenMPhillips** [comment]
Sudheesh, could you respond to the comments made by Jin Feng? If you have already discussed it with him in person, could you post a summary here.

I am in favor of merging this PR, but just the last commit. I think the second commit should be separate, and should be done as an inserted operator (similar to IteratorValidator), rather than modifying the constructor for Screen.

**Comment by sudheeshkatkam** [comment]
I have addressed @jinfengni 's and @hsuanyi 's comments [here](https://github.com/sudheeshkatkam/drill/commit/e4cfdfa9b0562d52ac07f6d80860a82fa8baba40) [I force pushed to this branch and somehow their comments are not referenced in this PR any longer.]

File: exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/limit/TestLimit0.java (hunk 2 above)
Comment (sudheeshkatkam) [inline exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/limit/TestLimit0.java]: All queries in this suite are optimized; I updated the PR.

**Comment by StevenMPhillips** [comment]
+1

**Comment by jinfengni** [comment]
LGTM.

+1

**Comment by sudheeshkatkam** [comment]
Thank you for the reviews.

All regression tests passed; I am running unit tests right now.

Note that, the `planner.enable_limit0_optimization` option is disabled by default. To summarize (and document) the limitations:

If, during validation, the planner is able to resolve that the types of the columns (i.e. types are non late binding), the shorter execution path is taken. Some types are excluded:
+ DECIMAL type is not fully supported in general.
+ VARBINARY is not fully tested.
+ MAP, ARRAY are currently not exposed to the planner.
+ TINYINT, SMALLINT are defined in the Drill type system but have been turned off for now.
+ SYMBOL, MULTISET, DISTINCT, STRUCTURED, ROW, OTHER, CURSOR, COLUMN_LIST are Calcite types currently not supported by Drill, nor defined in the Drill type list.

Three scenarios when the planner can do type resolution during validation:
+ Queries on Hive tables
+ Queries with explicit casts on table columns, example: `SELECT CAST(col1 AS BIGINT), ABS(CAST(col2 AS INTEGER)) FROM table;`
+ Queries on views with casts on table columns

In the latter two cases, the schema of the query with LIMIT 0 clause has relaxed nullability compared to the query without the LIMIT 0 clause. Example:
Say the schema definition of the Parquet file (`numbers.parquet`) is:
```
message Numbers {
  required int col1;
  optional int col2;
 }
```

Since the view definition does not specify nullability of columns, and schema of a parquet file is not yet leveraged by Drill's planner:
```
CREATE VIEW dfs.tmp.mynumbers AS SELECT CAST(col1 AS INTEGER) as col1, CAST(col2 AS INTEGER) AS col2 FROM dfs.tmp.`numbers.parquet`;
```
(1) For query with LIMIT 0 clause, since the file/ metadata is not read, Drill assumes the nullability of both columns is [`columnNullable`](https://docs.oracle.com/javase/7/docs/api/java/sql/ResultSetMetaData.html#columnNullable).
```
SELECT col1, col2 FROM dfs.tmp.mynumbers LIMIT 0;
```

(2) For query without LIMIT 0 clause, since the file is read, Drill knows the nullability of `col1` is [`columnNoNulls`](https://docs.oracle.com/javase/7/docs/api/java/sql/ResultSetMetaData.html#columnNoNulls), and `col2` is [`columnNullable`](https://docs.oracle.com/javase/7/docs/api/java/sql/ResultSetMetaData.html#columnNullable).
```
SELECT col1, col2 FROM dfs.tmp.mynumbers LIMIT 1;
```

**Pull request closed without merging**